### PR TITLE
Copy over TerminalWriter from py, simplify and optimize it

### PR DESCRIPTION
**Pull request by bluetech**
I was looking at some profiles, and noticed that when there is a lot of output, `TerminalWriter` features prominently. This prompted me to look into it.

`TerminalWriter` comes from `py`. `py` is frozen and needs to maintain backward compat and generality, so it's hard to do anything with that. So I decided to copy it over to pytest where it can be modified more freely. I believe pytest did something similar with the capturing infrastructure back in the day. And this also coincides with the goal of eventually removing the `py` dependency.

The commits in this PR detail every change I made starting from a straight copy. The main simplifications are:

- Remove all the Windows specific stuff -- my understanding is that coloroma already takes care of it, but I don't have Windows to test with.
- Remove functions unused by pytest.
- Remove Python2 stuff.
- Remove support for writing `bytes` which pytest doesn't need.
- Some general cleanups.

I also merged the changes from our custom `TerminalWriter` subclass, and ported the tests from py.

The file in `py` had 421, after all changes it is 180 lines not including the code from the subclass.

The optimizations are:

- Reduced overhead due to simplifications.
- Reduce amount of flushes to underlying file.
- Avoid expensive work of computing line widths when possible.

---

For a benchmark I used a semi-realistic scenario -- not a a micro-benchmark specifically of TerminalWriter, but something that does print a lot. Running the following with `pytest --collect-only`:

```py
import pytest
@pytest.mark.parametrize("x", range(50000))
def test_foo(x): pass
```

Before (leaving only relevant entries):

```
============================ no tests ran in 12.39s ============================
         23778937 function calls (22924066 primitive calls) in 12.674 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    50008    0.099    0.000    2.727    0.000 terminalwriter.py:270(line)
   100016    0.141    0.000    2.587    0.000 terminalwriter.py:245(write)
   100015    0.203    0.000    1.812    0.000 terminalwriter.py:258(_update_chars_on_current_line)
   100015    0.113    0.000    1.518    0.000 terminalwriter.py:72(get_line_width)
  1489288    0.710    0.000    1.125    0.000 terminalwriter.py:74(<genexpr>)
   100015    0.085    0.000    0.600    0.000 terminalwriter.py:399(write_out)
   100015    0.496    0.000    0.496    0.000 {method 'flush' of '_io.TextIOWrapper' objects}
```

After:

```
============================ no tests ran in 9.59s =============================
         17981953 function calls (17128590 primitive calls) in 9.741 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    50007    0.042    0.000    0.176    0.000 terminalwriter.py:175(line)
   100014    0.090    0.000    0.134    0.000 terminalwriter.py:159(write)
```

So at least in this scenario the overhead due to `TerminalWriter` is mostly eliminated.

### CODE REVIEW

File: src/_pytest/terminal.py [hunk 1]
```diff
@@ -359,8 +359,8 @@ def ensure_newline(self):
             self._tw.line()
             self.currentfspath = None
 
-    def write(self, content, **markup):
-        self._tw.write(content, **markup)
+    def write(self, content, flush: bool = False, **markup):
```
Comment (RonnyPfannschmidt) [inline src/_pytest/terminal.py:362]: lets ensure flush is a kwonly argument

File: src/_pytest/terminal.py [hunk 2]
```diff
@@ -539,24 +539,20 @@ def _write_progress_information_filling_space(self):
...
 
     @property
     def _width_of_current_line(self):
         """Return the width of current line, using the superior implementation of py-1.6 when available"""
-        try:
-            return self._tw.width_of_current_line
-        except AttributeError:
-            # py < 1.6.0
-            return self._tw.chars_on_current_line
+        return self._tw.width_of_current_line
```
Comment (RonnyPfannschmidt) [inline src/_pytest/terminal.py:547]: that one was waiting for removal since years, lovely

File: src/_pytest/_io/terminalwriter.py [hunk 3]
```diff
@@ -128,21 +136,43 @@ def ansi_print(text, esc, file=None, newline=True, flush=False):
     if flush:
         file.flush()
 
+
 def should_do_markup(file):
-    if os.environ.get('PY_COLORS') == '1':
+    if os.environ.get("PY_COLORS") == "1":
```
Comment (RonnyPfannschmidt) [inline src/_pytest/_io/terminalwriter.py:37]: whops, just noted, we should probably use `PYTEST_COLORS` here

**Comment by bluetech** [comment]
Thanks for the review @RonnyPfannschmidt!

I made the changes you requested, added a few small cleanups I noticed, and added a changelog detailing the breaking changes.

> we should probably use `PYTEST_COLORS` here

Unlike the other changes which *might* break some plugins, this one would be a breaking change for users. If we do decide to add a `PYTEST_COLORS`, it could subsume `PY_COLORS`, but would need to be documented properly. I'd be happy to do it, but prefer to do it in a separate PR.

**Comment by RonnyPfannschmidt** [comment]
Good call, this needs some work for configuration migration wrt py_colors vs a new name

**Comment by nicoddemus** [comment]
> as a followup i would like to note down an investigation for switching the kwarg based markup details with something more fluent

Definitely, that's something I've been thinking lately, after introducing Syntax Highlighting.

I've seen somewhere (I think in a `ward` issue/discussion) about [rich](https://pypi.org/project/rich). It supports inline markup. We could consider using the library directly or as inspiration.

But of course this is off-topic here. 😁

**Comment by nicoddemus** [comment]
I've tested this on Windows and didn't encounter any issues.

Great work @bluetech!

File: changelog/7135.breaking.rst [hunk 4]
```diff
@@ -0,0 +1,15 @@
+Pytest now uses its own ``TerminalWriter`` class instead of using the one from the ``py`` library.
```
Comment (nicoddemus) [inline changelog/7135.breaking.rst:1]: Great changelog. 👍

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus a few, years ago I tried to do terminalwriter in terms of click

**Comment by nicoddemus** [comment]
Thanks @bluetech!

**Pull request merged**